django/db/models/sql/compiler.py

```diff
@@ -903,17 +905,102 @@ def __init__(self, *args, **kwargs):
         self.return_id = False
         super(SQLInsertCompiler, self).__init__(*args, **kwargs)
 
-    def placeholder(self, field, val):
+    def field_as_sql(self, field, val):
+        """
+        Take a field, and a value intended to be saved on that field, and
+        return placeholder SQL and accompanying params. Checks for raw values,
+        expressions and fields with get_placeholder() defined in that order.
+        """
+
         if field is None:
             # A field value of None means the value is raw.
-            return val
+            sql, params = val, []
+
+        elif hasattr(val, 'as_sql'):
+            # This is an expression, let's compile it.
+            sql, params = self.compile(val)
+
         elif hasattr(field, 'get_placeholder'):
             # Some fields (e.g. geo fields) need special munging before
             # they can be inserted.
-            return field.get_placeholder(val, self, self.connection)
+            sql, params = field.get_placeholder(val, self, self.connection), [val]
+
         else:
             # Return the common case for the placeholder
-            return '%s'
+            sql, params = '%s', [val]
+
+        # The following hook is only used by Oracle Spatial, which sometimes
+        # needs to yield 'NULL' and [] as its placeholder and params instead
+        # of '%s' and [None]. The 'NULL' placeholder is produced earlier by
+        # OracleOperations.get_geom_placeholder(). The following line removes
+        # the corresponding None parameter. See ticket #10888.
+        params = self.connection.ops.modify_insert_params(sql, params)
+
+        return sql, params
+
+    def prepare_value(self, field, value):
+        """
+        Prepare a value to be used in a query, by resolving it if it is an
+        expression, and otherwise calling the field's get_db_prep_save().
+        """
+
+        def find_fs(expr):
+            """Walk this expression tree depth-first and yield all F nodes."""
+            if isinstance(expr, F):
+                yield expr
+            elif isinstance(expr, CombinedExpression):
+                for e in chain(find_fs(expr.lhs), find_fs(expr.rhs)):
+                    yield e
+
+        if hasattr(value, 'resolve_expression'):
+            # Don't allow F() expressions. They refer to existing columns on a
+            # row, but in the case of insert the row doesn't exist yet.
+            if any(find_fs(value)):
+                raise ValueError(
+                    'Failed to insert expression "%s" on %s. F() expressions '
+                    'can only be used to update, not to insert.' % (value, field)
+                )
```

Comment:
F() expressions aren't the only ones that can refer to other columns in the query. How about Q(other_field__isnull=True). Also, expressions are free to resolve columns of the query without using F-expressions. We need some other way to know if the expression refers to columns of the query.

Maybe we could first resolve the expression, the check for Col references? That might be better.

The check should be done in the expression, so that the expression tells Django if it is referring to any columns. Making the compiler guess this is the wrong way in my opinion.